### PR TITLE
Implement source cmds 2 repo context

### PR DESCRIPTION
**Pull request by ankit-r-gajera**
_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
_Provide a 1-2 line brief overview of the changes submitted through the Pull Request..._

### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): **\<URL to be filled>**
- [ ] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [ ] Added label to the Pull Request for easier discoverability and search
- [ ] RTC or HSD number will be included in final merge. HSD must always be included if available.
- [ ] Changelogs are updated (or N/A if not customer visible)
- [ ] inbm/log_changes.txt and inbm-vision/log_changes.txt are updated for potentially Validation-breaking log changes (or N/A if none)

### CODE MAINTAINABILITY
- [ ] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [ ] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [ ] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [ ] Updated Documentation as relevant to the changes
- [ ] Updated Build steps/commands changes as relevant
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- [ ] Specific instructions or information for code reviewers (If any):
- [ ] Run 'go fmt' or format-python.sh as applicable.
- [ ] New/modified methods and functions should have type annotations on signatures as applicable
- [ ] New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- [ ] Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- [ ] Is cloud UI changed? If so, are cloud definition files updated?

Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [ ] **_Quality of code (At least one should be checked as applicable)\*_**
	- [ ] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [ ] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [ ] **_Static code scan report shows zero critical issues\*_**
- [ ] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

### CODE REVIEW

**Comment by ankit-r-gajera** [comment]
Preparing review...

**Comment by ankit-r-gajera** [comment]
Preparing review...

**Comment by ankit-r-gajera** [comment]
Preparing review...

**Comment by ankit-r-gajera** [comment]
## Code Review Analysis

- 🎯 **Main theme:** Implementing source commands for repository context
- 📝 **PR summary:** This PR introduces the ability to manage OS and application sources for package repositories. It includes commands to add, remove, update, and list sources for both OS and applications. The changes are primarily focused on the Ubuntu operating system.
- 📌 **Type of PR:** Enhancement
- 🏅 **Score:** 89
- 🧪 **Relevant tests added:** No
- ⏱️ **Estimated effort to review [1-5]:** 3, because the PR includes a variety of new features and changes that require careful review to ensure they are implemented correctly and securely.
- 🔒 **Security concerns:** No
## Code Review Feedback



<strong>💡 General suggestions:</strong> The PR introduces a significant feature to the manageability framework, allowing users to manage package sources for OS and applications. It is important to ensure that the new commands are secure, especially when dealing with GPG keys and repository sources. Additionally, it would be beneficial to have integration tests that cover the new functionality to ensure it works as expected and does not introduce regressions.


<details> <summary><strong>✨ Usage tips:</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 

</details>

File: inbm/dispatcher-agent/fpm-template/usr/share/dispatcher-agent/manifest_schema.xsd [hunk 1]
```diff
@@ -322,16 +322,22 @@
...
                                         </xs:element>
                                         <xs:element name="repo" minOccurs="1" maxOccurs="1">
                                             <xs:complexType>
                                                 <xs:sequence>
-                                                    <xs:element name="source" type="Max200Chars" minOccurs="0" maxOccurs="1"/>
-                                                    <xs:element name="filename" type="Max200Chars" minOccurs="0" maxOccurs="1"/>
+                                                    <xs:element name="repos" minOccurs="1" maxOccurs="1">
+                                                        <xs:complexType>
+                                                            <xs:sequence>
+                                                                <xs:element name="source_pkg" type="Max200Chars" minOccurs="1" maxOccurs="2"/>
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/usr/share/dispatcher-agent/manifest_schema.xsd:336]: Add validation for the new `<source_pkg>` elements to ensure they contain valid repository source strings. [important]

File: inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher [hunk 2]
```diff
@@ -59,7 +59,8 @@
   /etc/apt/sources.list.bak rw,
   /etc/apt/sources.list.d/* rw,
   /etc/apt/sources.list.d/ rw,
-  /usr/bin/gpg px,
+  /usr/bin/gpg rUx,
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher:62]: Ensure that the AppArmor profile allows the dispatcher to perform all necessary operations with GPG, such as importing keys and verifying signatures. [important]

File: inbm/dispatcher-agent/pyproject.toml [hunk 3]
```diff
@@ -40,7 +40,7 @@ max_tuple_unpack_length = 4
 
 [tool.flakeheaven.plugins]
 "flake8-annotations" = ["-*", "+ANN001", "+ANN2??", "+ANN301"]
-"wemake-python-styleguide" = ["-*", "+WPS2??", "-WPS220", "-WPS223", "-WPS222", "-WPS236", "-WPS237", "-WPS238"] # WPS2?? is complexity; 220 is deep nesting; 223 is elifs; 222 is logic; 236 is # of variables unpacked in a tuple; 237 is complex format string; 238 is # of raises in a function
+"wemake-python-styleguide" = ["-*", "+WPS2??", "-WPS220", "-WPS223", "-WPS222", "-WPS225", "-WPS236", "-WPS237", "-WPS238"] # WPS2?? is complexity; 220 is deep nesting; 223 is elifs; 222 is logic; 236 is # of variables unpacked in a tuple; 237 is complex format string; 238 is # of raises in a function
 "flake8-bandit" = ["+*", "-S404", "-S603"]
 
 [tool.pytest.ini_options]
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/pyproject.toml:46]: If new dependencies were added for handling source commands, ensure they are included in the project's dependency list. [medium]

File: inbm/dispatcher-agent/fpm-template/usr/share/dispatcher-agent/manifest_schema.xsd (hunk 1 above)
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/usr/share/dispatcher-agent/manifest_schema.xsd:336]: Consider adding a pattern restriction to the `<source_pkg>` element to enforce proper repository format. [medium]

File: inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher [hunk 4]
```diff
@@ -151,6 +152,7 @@
   /usr/share/configuration-agent/iotg_inb_schema.xsd l,
   /usr/share/configuration-agent/iotg_inb_schema.xsd.dpkg-tmp l,
   /usr/share/diagnostic-agent/ w,
+  /usr/share/keyrings/*.gpg w,
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher:155]: Review the AppArmor profile to ensure that it does not grant overly permissive access to the dispatcher-agent, especially in terms of file write permissions. [important]

**Comment by ankit-r-gajera** [comment]
Preparing code suggestions...

**Comment by ankit-r-gajera** [comment]
Preparing review...

**Comment by ankit-r-gajera** [comment]
## Code Review Analysis

- 🎯 **Main theme:** Implementing source commands for repository context
- 📝 **PR summary:** This PR introduces the ability to manage OS and application sources for package repositories. It includes commands to add, remove, update, and list sources for both OS and application package repositories. These commands are essential for maintaining the list of package sources that the system uses to fetch updates.
- 📌 **Type of PR:** Enhancement
- 🏅 **Score:** 89
- 🧪 **Relevant tests added:** No
- ⏱️ **Estimated effort to review [1-5]:** 3, because the PR includes multiple new features and changes to existing code, which require careful review to ensure they work as intended and do not introduce regressions.
- 🔒 **Security concerns:** No
## Code Review Feedback



<strong>💡 General suggestions:</strong> The PR introduces a set of new commands for managing package sources, which is a valuable addition to the system's capabilities. It is important to ensure that these commands are well-tested, especially since the PR does not include unit tests. The maintainers should consider adding relevant tests to cover the new functionality. Additionally, it would be beneficial to ensure that the new commands are consistent with the system's security policies, particularly with regards to handling GPG keys and modifying source lists.


<details> <summary><strong>✨ Usage tips:</strong></summary><hr> 

> To invoke the PR-Agent, add a comment using one of the following commands:
> **/review [-i]**: Request a review of your Pull Request. For an incremental review, which only considers changes since the last review, include the '-i' option.
> **/describe**: Modify the PR title and description based on the contents of the PR.
>For example: /review --pr_reviewer.extra_instructions="focus on the file: ..." 

</details>

File: inbm/dispatcher-agent/fpm-template/usr/share/dispatcher-agent/manifest_schema.xsd (hunk 1 above)
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/usr/share/dispatcher-agent/manifest_schema.xsd:336]: Consider adding a `minOccurs="0"` attribute to the `<source_pkg>` element within the `<repos>` complexType for both the `<add>` and `<update>` elements. This change would allow for the possibility of having an empty list of sources, which might be a valid case in certain scenarios. [medium]

Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/usr/share/dispatcher-agent/manifest_schema.xsd:336]: The `maxOccurs` attribute for `<source_pkg>` within the `<repos>` complexType is set to "2", which seems to limit the number of sources that can be added or updated to two. If the intention is to allow more than two sources, this value should be increased or set to "unbounded". [important]

File: inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher [hunk 5]
```diff
@@ -178,6 +180,8 @@
   /usr/share/trtl/trtl.xsd r,
   /usr/share/trtl/ w,
   /usr/share/trtl/* rw,
+  /usr/share/keyrings/ rw,
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher:183]: The AppArmor profile should explicitly deny writing to the `/usr/share/keyrings/` directory to prevent unauthorized modifications to the GPG keyrings. [important]

File: inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher (hunk 4 above)
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/fpm-template/etc/apparmor.d/usr.bin.inbm-dispatcher:155]: The AppArmor profile should restrict write access to specific keyring files instead of allowing write access to all files in `/usr/share/keyrings/`. Use a more specific pattern to match only the necessary keyring files. [medium]

File: inbm/dispatcher-agent/pyproject.toml [hunk 6]
```diff
@@ -40,7 +40,7 @@ max_tuple_unpack_length = 4
 
 [tool.flakeheaven.plugins]
 "flake8-annotations" = ["-*", "+ANN001", "+ANN2??", "+ANN301"]
-"wemake-python-styleguide" = ["-*", "+WPS2??", "-WPS220", "-WPS223", "-WPS222", "-WPS236", "-WPS237", "-WPS238"] # WPS2?? is complexity; 220 is deep nesting; 223 is elifs; 222 is logic; 236 is # of variables unpacked in a tuple; 237 is complex format string; 238 is # of raises in a function
+"wemake-python-styleguide" = ["-*", "+WPS2??", "-WPS220", "-WPS223", "-WPS222", "-WPS225", "-WPS236", "-WPS237", "-WPS238"] # WPS2?? is complexity; 220 is deep nesting; 223 is elifs; 222 is logic; 236 is # of variables unpacked in a tuple; 237 is complex format string; 238 is # of raises in a function
 "flake8-bandit" = ["+*", "-S404", "-S603"]
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/pyproject.toml:44]: Ensure that the code style checks and linter configurations in `pyproject.toml` are consistent with the new code added in the PR. If new rules or exceptions are needed, update the configuration accordingly. [medium]

File: inbm/dispatcher-agent/tests/unit/source/test_ubuntu_source_cmd.py [hunk 7]
```diff
@@ -111,14 +119,100 @@ def write_side_effect(*args, **kwargs):
...
+    # def test_raises_exception_on_io_error_during_update_app_source_ubuntu(self):
+    #     params = ApplicationUpdateSourceParameters(file_name="intel-gpu-jammy.list",
+    #                                                sources=APP_SOURCE)
+    #     command = UbuntuApplicationSourceManager()
+    #     with pytest.raises(SourceError) as exc_info:
+    #         with patch('builtins.open', new_callable=mock_open(), side_effect=OSError):
+    #             command.update(params)
+    #     assert "Error while writing file: " in str(exc_info.value)
+
     def test_list(self, sources_list_d_content):
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/tests/unit/source/test_ubuntu_source_cmd.py:216]: Add unit tests for the new source commands to ensure they behave as expected. This includes testing the success paths as well as error handling and edge cases. [important]

File: inbm/dispatcher-agent/tests/unit/test_xmlparser.py [hunk 8]
```diff
@@ -103,11 +103,13 @@ def test_application_source_with_add_passes_validation(self) -> None:
```
Comment (ankit-r-gajera) [inline inbm/dispatcher-agent/tests/unit/test_xmlparser.py:102]: Update the unit tests for XML parsing to cover the new source command XML structures. This will help catch any issues with the manifest schema and ensure that the XML commands are parsed correctly. [important]

**Comment by ankit-r-gajera** [comment]
Preparing code suggestions...
